### PR TITLE
Allow limiting watched namespaces

### PR DESCRIPTION
**Pull request by bison**
This uses `MultiNamespaceListerWatcher` to allow Istiod and Mixer to watch a set of namespaces, instead of being limited to either a single namespace or all namespaces.

### CODE REVIEW

**Comment by bison** [comment]
This doesn't touch the webhooks. Not 100% sure what the best approach would be there. The webhook server could filter each request, but that doesn't seem ideal. Istiod could also label namespaces with an ID, and use that in the webhook's namespace selector. I suppose it would have to do some kind of clean up if the list changes in that case though.

**Comment by rcernich** [comment]
@bison, configuration of watched namespaces should be a separate discussion.  This PR is a good starting point, as it facilitates configuration of specific watched namespaces vs a cluster wide watch, and lays the foundation for a more constrained controller implementation.

**Comment by bison** [comment]
>  configuration of watched namespaces should be a separate discussion. This PR is a good starting point, as it facilitates configuration of specific watched namespaces vs a cluster wide watch, and lays the foundation for a more constrained controller implementation.

@rcernich, definitely agree. Not sure if that was in response to my comment. My only concern with this right now is that the validating and injection webhooks aren't limited to the same set of namespaces, and they need to be. Need to figure out how to do that.

**Comment by rcernich** [comment]
Thanks for the clarification @bison.  I think that's okay for now.  As I said, this provides a good foundation for configuring a more constrained mesh.

**Comment by bison** [comment]
@dgn, thanks. Yeah, absolutely. I updated `TestPodCache` to that end, and I'll work on some more unit tests.

File: pilot/pkg/bootstrap/server.go [hunk 1]
```diff
@@ -179,6 +179,14 @@ func NewServer(args *PilotArgs) (*Server, error) {
 		mux:            http.NewServeMux(),
 	}
 
+	if args.Config.ControllerOptions.WatchedNamespaces != "" {
+		// Add the control-plane namespace to the list of watched namespaces.
```
Comment (howardjohn) [inline pilot/pkg/bootstrap/server.go:183]: Why do we need this? It seems reasonable just curious as it doesn't seem strictly required. I guess we need it to read the secrets, etc that reside in the same ns?

Comment (dgn) [reply pilot/pkg/bootstrap/server.go:183]: IIRC we have some CRs that behave differently when deployed to the 'root namespace' (they might have been deprecated by now though?), plus by default we deploy the gateways to the same namespace as istiod

Comment (bison) [reply pilot/pkg/bootstrap/server.go:183]: That's the primary reason, yeah. Listing things like secrets and pods (ingress / egress gateways for instance) in the control-plane namespace. Probably doable to add the control-plane namespace directly to the pod cache or when needed for secrets listers etc., but that seemed a little error prone.

Comment (dgn) [reply pilot/pkg/bootstrap/server.go:183]: regarding the gateways this might actually be argument to append the namespaces at install time, though

Comment (costinm) [reply pilot/pkg/bootstrap/server.go:183]: I would remove this code - if user wants explicit control over watched namespaces, they should include the control plane ns. If they don't - they shouldn't. 

istio-system is messy, in particular with multiple revisions running. In many cases it's safer to set 'istio-config' to a different namespace or not use it at all. 

I wouldn't even add istio-system explicitly in installer either - let user do it.

File: pilot/cmd/pilot-discovery/main.go [hunk 2]
```diff
@@ -144,9 +144,8 @@ func init() {
 		"Setting this flag has no effect. Install CRD definitions directly or with the operator")
 	discoveryCmd.PersistentFlags().StringVar(&serverArgs.Config.FileDir, "configDir", "",
 		"Directory to watch for updates to config yaml files. If specified, the files will be used as the source of config, rather than a CRD client.")
-	discoveryCmd.PersistentFlags().StringVarP(&serverArgs.Config.ControllerOptions.WatchedNamespace, "appNamespace",
-		"a", metav1.NamespaceAll,
-		"Restrict the applications namespace the controller manages; if not set, controller watches all namespaces")
+	discoveryCmd.PersistentFlags().StringVarP(&serverArgs.Config.ControllerOptions.WatchedNamespaces, "appNamespace", "a", metav1.NamespaceAll,
```
Comment (costinm) [inline pilot/cmd/pilot-discovery/main.go:147]: Please deprecate this option. Use MeshConfig !

**Comment by pnambiarsf** [comment]
@linsun @rshriram - Can this be merged today ? We (Salesforce) is blocked by this. Looks like it needs approval from two WGs. Who should I tag here ?

File: pilot/cmd/pilot-discovery/main.go (hunk 2 above)
Comment (bison) [reply pilot/cmd/pilot-discovery/main.go:147]: Planning on doing that next. Just didn't want to make this PR any larger than necessary.

**Comment by bison** [comment]
> We should add a global option for this in API so that users don't have to change the watched namespaces via flags at multiple places.

Yeah, the list of namespaces should just go into MeshConfig. Was planning on doing that next, but wanted to avoid making this PR any larger than necessary.

**Comment by sdake** [comment]
@bison nice first commit to Istio! Hope to see more great work like this.

Cheers
-steve

**Pull request merged**